Sources/NIOPosix/NonBlockingFileIO.swift

```diff
@@ -712,3 +726,329 @@ public struct NIODirectoryEntry: Hashable {
+    /// - parameters:
+    ///   - fileHandle: The `NIOFileHandle` to read from.
+    public func readFileSize(fileHandle: NIOFileHandle) async throws -> Int64 {
+        return try await self.threadPool.runIfActive {
+            return try fileHandle.withUnsafeFileDescriptor { descriptor in
+                let curr = try Posix.lseek(descriptor: descriptor, offset: 0, whence: SEEK_CUR)
+                let eof = try Posix.lseek(descriptor: descriptor, offset: 0, whence: SEEK_END)
+                try Posix.lseek(descriptor: descriptor, offset: curr, whence: SEEK_SET)
+                return Int64(eof)
+            }
+        }
+    }
+
+    /// Write `buffer` to `fileHandle` in `NonBlockingFileIO`'s private thread pool which is separate from any `EventLoop` thread.
+    ///
+    /// - parameters:
+    ///   - fileHandle: The `NIOFileHandle` to write to.
+    ///   - buffer: The `ByteBuffer` to write.
+    public func write(fileHandle: NIOFileHandle,
+                      buffer: ByteBuffer) async throws {
+        return try await self.write0(fileHandle: fileHandle, toOffset: nil, buffer: buffer)
+    }
+
+    /// Write `buffer` starting from `toOffset` to `fileHandle` in `NonBlockingFileIO`'s private thread pool which is separate from any `EventLoop` thread.
+    ///
+    /// - parameters:
+    ///   - fileHandle: The `NIOFileHandle` to write to.
+    ///   - toOffset: The file offset to write to.
+    ///   - buffer: The `ByteBuffer` to write.
+    public func write(fileHandle: NIOFileHandle,
+                      toOffset: Int64,
+                      buffer: ByteBuffer) async throws {
+        return try await self.write0(fileHandle: fileHandle, toOffset: toOffset, buffer: buffer)
+    }
+
+    private func write0(fileHandle: NIOFileHandle,
+                        toOffset: Int64?,
+                        buffer: ByteBuffer) async throws {
+        let byteCount = buffer.readableBytes
+
+        guard byteCount > 0 else {
+            return
+        }
+
+        return try await self.threadPool.runIfActive {
+            try writeSync(fileHandle: fileHandle, byteCount: byteCount, toOffset: toOffset, buffer: buffer)
+        }
+    }
+
+    /// Open the file at `path` for reading on a private thread pool which is separate from any `EventLoop` thread.
+    ///
+    /// This function will return (a future) of the `NIOFileHandle` associated with the file opened and a `FileRegion`
+    /// comprising of the whole file. The caller must close the returned `NIOFileHandle` when it's no longer needed.
+    ///
+    /// - note: The reason this returns the `NIOFileHandle` and the `FileRegion` is that both the opening of a file as well as the querying of its size are blocking.
+    ///
+    /// - parameters:
+    ///     - path: The path of the file to be opened for reading.
+    /// - returns: the `NIOFileHandle` and the `FileRegion` comprising the whole file.
+    public func openFile(path: String) async throws -> (NIOFileHandle, FileRegion) {
```

Comment:
We should probably spell this as `func withOpenFile<Result>(path: String, _ body: (NIOFileHandle, FileRegion) async throws -> Result) async throws -> Result`. This allows us to keep the file handle open for the duration of the `with` body and we can close at the end providing the user an even simpler API where they don't have to take care of the closing explicitly.

Reply:
We also need to be careful to make sure we close _even if_ the task has been cancelled.

Reply:
Also while `NIOFileHandle` isn't `Sendable` I'm wondering if we should at least make it thread safe. At the moment a double close could result in closing someone else's fd, we can and I think should, avoid doing that.

Reply:
`withOpenFile` is obviously different from the ELF version of this function. But I'm open to making this change.

Reply:
True, I think it's okay to diverge here given how easy it is to not `close()` given the lack of an `async` `defer`.

Reply:
I was just caught out by this right now 🤣

Reply:
done